lib/OptimizationMOI/test/runtests.jl

```diff
@@ -40,4 +40,12 @@ using Test
 
     sol = solve(prob, OptimizationMOI.MOI.OptimizerWithAttributes(Ipopt.Optimizer, "max_cpu_time" => 60.0))
     @test 10 * sol.minimum < l1
+
+    nl_solver = OptimizationMOI.MOI.OptimizerWithAttributes(Ipopt.Optimizer, "print_level" => 0)
+    minlp_solver = OptimizationMOI.MOI.OptimizerWithAttributes(Juniper.Optimizer, "nl_solver" => nl_solver)
+
+    sol = solve(prob, minlp_solver)
+    @test 10 * sol.minimum < l1
+
+    sol = solve(prob, OptimizationMOI.MOI.OptimizerWithAttributes( () -> AmplNLWriter.Optimizer(Ipopt_jll.amplexe)))
```

Comment:
@odow this throws 

```
ERROR: MathOptInterface.AddVariableNotAllowed: Adding variables cannot be performed. You may want to use a `CachingOptimizer` in `AUTOMATIC` mode or you may need to call `reset_optimizer` before doing this operation if the `CachingOptimizer` is in `MANUAL` mode.
Stacktrace:
 [1] add_variable(model::AmplNLWriter.Optimizer)
   @ MathOptInterface ~/.julia/packages/MathOptInterface/RuRWI/src/variables.jl:43
 [2] add_variables(model::AmplNLWriter.Optimizer, n::Int64)
   @ MathOptInterface ~/.julia/packages/MathOptInterface/RuRWI/src/variables.jl:32
 [3] __solve(prob::OptimizationProblem{true, OptimizationFunction{true, Optimization.AutoModelingToolkit, typeof(rosenbrock), Nothing, Nothing, Nothing, var"#11#12", Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing}, Vector{Float64}, Vector{Float64}, Nothing, Vector{Float64}, Vector{Float64}, Nothing, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}}, opt::MathOptInterface.OptimizerWithAttributes; maxiters::Nothing, maxtime::Nothing, abstol::Nothing, reltol::Nothing, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ OptimizationMOI ~/Optimization.jl/lib/OptimizationMOI/src/OptimizationMOI.jl:225
```

Reply:
Ah. This is a little complicated. MathOptInterface has two types of solvers: incremental and copy-to. Incremental solvers support building the problem piece by piece, and copy-to need it all in one go. Ipopt supports incremental building, but AmplNLWriter doesn't.

You have a couple of options.

1) Change
```Julia
https://github.com/SciML/Optimization.jl/blob/ad41c6e7c870dbdd1785d577640a96ad671ca00d/lib/OptimizationMOI/src/OptimizationMOI.jl#L159
```
to 
```julia
_create_new_optimizer(opt) = MOI.instantiate(opt; with_bridge_type = Float64)
```
(Note that I dropped the type annotation.)


2) Use a caching optimizer, but that's not very nice for users:
```julia
() - > MOI.Utilities.CachingOptimizer(
    MOI.Utilities.UniversalFallback(MOI.Utilities.Model{Float64}()),
    AmplNLWriter.Optimizer(Ipopt_jll.amplexe),
)
```

3) explicitly pass:
```julia
MOI.instantiate(() -> AmplNLWriter.Optimizer(Ipopt_jll.amplexe); with_bridge_type = Float64)
```

Ideally, you wouldn't have this method, 
https://github.com/SciML/Optimization.jl/blob/ad41c6e7c870dbdd1785d577640a96ad671ca00d/lib/OptimizationMOI/src/OptimizationMOI.jl#L158
and people would go `solve(prob, Ipopt.Optimizer)` instead of  `solve(prob, Ipopt.Optimizer())`.

Reply:
Are `x₁` the variables? You also need to pass `Expr(:ref, :x, x₁)` or `:(x[$x₁])`

Reply:
Yes, thanks, I was facing that now and trying to figure it out 😅

Reply:
```
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol +
    2: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol *
        2: Float64 100.0
        3: Expr
          head: Symbol call
          args: Array{Any}((3,))
            1: Symbol ^
            2: Expr
              head: Symbol call
              args: Array{Any}((3,))
                1: Symbol +
                2: Expr
                3: Expr
            3: Int64 2
    3: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol ^
        2: Expr
          head: Symbol call
          args: Array{Any}((3,))
            1: Symbol +
            2: Float64 1.0
            3: Expr
              head: Symbol call
              args: Array{Any}((3,))
                1: Symbol *
                2: Int64 -1
                3: Expr
        3: Int64 2
nothing
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ==
    2: Expr
      head: Symbol call
      args: Array{Any}((3,))
        1: Symbol +
        2: Expr
          head: Symbol call
          args: Array{Any}((3,))
            1: Symbol ^
            2: Expr
              head: Symbol ref
              args: Array{Any}((2,))
                1: Symbol x
                2: Int64 1
            3: Int64 2
        3: Expr
          head: Symbol call
          args: Array{Any}((3,))
            1: Symbol ^
            2: Expr
              head: Symbol ref
              args: Array{Any}((2,))
                1: Symbol x
                2: Int64 2
            3: Int64 2
    3: Int64 0
[nothing]
Ipopt 3.14.4: 
bad line 49 of /var/folders/p2/0zng1mgn5gq3j99h7rflx1n40000gn/T/jl_iPDTXW/model.nl: J0 0
Cannot open .nl file
libc++abi: terminating with uncaught exception of type Ipopt::TNLP::INVALID_TNLP
u: 2-element Vector{Float64}:
 NaN
 NaN
 ```
 
 the contents of `model.nl` are
 
 ```
 g3 1 1 0
 2 1 1 0 1 0
 1 1
 0 0
 0 0 0
 0 0 0 1
 0 0 0 0 0
 0 0
 0 0
 0 0 0 0 0
C0
o0
o5
n1
n2
o5
n2
n2
O0 0
o0
o2
n100
o5
o0
n2
o2
n-1
o5
n1
n2
n2
o5
o0
n1
o2
n-1
n1
n2
x2
0 0
1 0
r
4 0
b
3
3
k1
0
J0 0
```

Reply:
This is fixed. We needed to pass `x[MOI.VariableIndex(1)]` instead of `x[1]`.

Reply:
Thanks you!